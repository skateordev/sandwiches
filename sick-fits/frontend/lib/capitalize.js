/**
 * Takes the given string and makes the first character uppercase
 * @param string: the string to capitalize
 */

function capitalize(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

export default capitalize;
