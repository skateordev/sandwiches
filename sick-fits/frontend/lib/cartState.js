import PropTypes from 'prop-types';
import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // this is our own custom provider!
  // we will store data (state) and functionality (updaters) in here
  // anyone can acces it via the consumer!

  const [cartOpen, setCartOpen] = useState(false);

  function openCart() {
    setCartOpen(true);
  }

  function closeCart() {
    setCartOpen(false);
  }

  return (
    <LocalStateProvider value={{ cartOpen, openCart, closeCart }}>
      {children}
    </LocalStateProvider>
  );
}

CartStateProvider.propTypes = {
  children: PropTypes.node,
};

CartStateProvider.defaultProps = {
  children: null,
};

// make a custom hook for access the cart local state
function useCart() {
  // use a consumer to access the local state
  const all = useContext(LocalStateContext);

  return all;
}

export {
  useCart,
  CartStateProvider,
};
