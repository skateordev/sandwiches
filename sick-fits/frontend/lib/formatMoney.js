export default function formatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'EUR',
    minimumFractionDigits: 2,
  };

  // check if it's a whole dollar amount
  if (amount % 100 === 0) {
    options.minimumFractionDigits = 0;
  }

  // english-Eurozone (countries that fully support the euro)
  const formatter = Intl.NumberFormat('en-EZ', options);

  return formatter.format(amount / 100);
}
