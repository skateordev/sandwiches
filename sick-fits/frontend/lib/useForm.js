import { useEffect, useState } from "react"

export default function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial);

  // fix values of `initial` being undefined
  const initialValues = Object.values(initial).join('');

  useEffect(() => {
    // initialValues will be undefined on first render
    // update inputs to the actual values once initialValues changes
    setInputs(initial);
  }, [initialValues]);

  function handleChange(evt) {
    let { name, type, files, value } = evt.target;

    if (type === 'number') {
      value = parseInt(value);
    }

    if (type === 'file') {
      // assign files to the first item (aka first file) in the [value] array
      [value] = files;
    }

    setInputs({
      ...inputs,
      [name]: value,
    })
  }

  function resetForm() {
    setInputs(initial)
  }

  function clearForm() {
    console.log('inputs', inputs);
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );

    setInputs(blankState);
  }

  return {
    inputs,
    clearForm,
    resetForm,
    handleChange,
  }
};
