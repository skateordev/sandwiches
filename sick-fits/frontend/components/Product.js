import Link from 'next/link';
import formatMoney from '../lib/formatMoney';
import { ItemStyled, TitleStyled, PriceTagStyled } from './styles';


export default function Product({
  product
}) {
  const {
    id,
    name,
    photo,
    price,
    description,
  } = product;

  return (
    <ItemStyled>
      <img src={photo?.image?.publicUrlTransformed} alt={name} />
      <TitleStyled>
        <Link href={`/product/${id}`}>{name}</Link>
      </TitleStyled>
      <PriceTagStyled>{formatMoney(price)}</PriceTagStyled>
      <p>{description}</p>
      {/* Add buttons to edit and delete */}
    </ItemStyled>
  )
};
