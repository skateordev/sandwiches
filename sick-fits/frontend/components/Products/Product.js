import PropTypes from 'prop-types';
import Link from 'next/link';
import formatMoney from '../../lib/formatMoney';
import { ItemStyled, TitleStyled, PriceTagStyled } from '../styles';
import AddToCart from '../AddToCart/AddToCart';

export default function Product({
  product,
}) {
  const {
    id: productId,
    name,
    photo,
    price,
    description,
  } = product;

  return (
    <ItemStyled>
      <img src={photo?.image?.publicUrlTransformed} alt={name} />
      <TitleStyled>
        <Link href={`/product/${productId}`}>{name}</Link>
      </TitleStyled>
      <PriceTagStyled>{formatMoney(price)}</PriceTagStyled>
      <p>{description}</p>
      {/* Add buttons to edit and delete */}
      <div className="buttonList">
        <Link href={{
          pathname: '/update',
          query: { id: productId },
        }}
        >
          Edit üìù
        </Link>
        <AddToCart productId={productId} />
      </div>
    </ItemStyled>
  );
}

Product.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    photo: PropTypes.shape({
      id: PropTypes.string,
      image: PropTypes.shape({
        publicUrlTransformed: PropTypes.string,
      }),
    }),
    price: PropTypes.number,
    description: PropTypes.string,
  }).isRequired,
};
