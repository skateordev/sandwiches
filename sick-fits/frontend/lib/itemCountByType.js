/**
 * Determines the number of items for the given itemType
 * Used primarily for pagination and cache
 * @param {object} data: object read from cache or returnd from paginationQuery
 * @param {string} itemType
 * @returns {number} itemType meta count property
 */
import { ITEM_TYPE_ORDERS, ITEM_TYPE_PRODUCTS } from '../components/constants';

export default function itemCountByType(data, itemType) {
  let count = 0;

  switch (itemType) {
    case ITEM_TYPE_ORDERS:
      count = data?._allOrdersMeta?.count;
      break;
    case ITEM_TYPE_PRODUCTS:
      count = data?._allProductsMeta?.count;
      break;
    default:
      break;
  }

  return count;
}
