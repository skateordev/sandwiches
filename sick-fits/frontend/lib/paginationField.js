import { ITEM_TYPE_ORDERS, ITEM_TYPE_PRODUCTS } from '../components/constants';

export default function paginationField(query, itemType) {
  return ({
    keyArgs: false, // tells apollo we will take care of everything
    read: (existing = [], { args, cache }) => { // eslint-disable-line default-param-last
      /**  first we ask `read` function for some items
        *  we can either:
        *    return items already in cache
        *    - or -
        *    return false (and make/cause a network request to refresh) */

      const { skip, first } = args;

      // read number of items on the page from cache
      const data = cache.readQuery({ query });
      let count = 0;

      // determine what we're paginating
      switch (itemType) {
        case ITEM_TYPE_ORDERS:
          count = data?._allOrdersMeta?.count;
          break;
        case ITEM_TYPE_PRODUCTS:
          count = data?._allProductsMeta?.count;
          break;
        default:
          break;
      }

      // current page number and total number of pages
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      /* check if we have existing items and also filter out undefined
         if doing 4 items per page then on the 2nd page skip would be 4
         skip + first = 8, so items = 5, 6, 7, 8 */
      const items = existing.slice(skip, skip + first).filter((item) => item);

      if (items.length && items.length !== first && page === pages) {
        /* when there aren't enough items for a full page (partial last page)
           just gimme whatever there is */

        return items;
      }

      if (items.length !== first) {
        // there are no items! go to network and fetch!

        return false;
      }

      if (items.length) {
        /* items have been found in cache, don't need to
           go to network */
        console.log(`💥 GOT EEEM!\n🔍 Found ${existing.length} cached items!\n🪂 Sending to apollo...`);

        return items;
      }

      return false; // fallback to network request
    },
    merge: (existing, incoming, { args }) => {
      // this runs when the Apollo client comes back from the network with our product
      const { skip } = args;

      console.log(`Merging ${incoming.length} items from the network 🌐`);

      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; i++) {
        merged[i] = incoming[i - skip];
      }

      return merged;
    },
  });
}
