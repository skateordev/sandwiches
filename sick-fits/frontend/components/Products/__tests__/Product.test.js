/* eslint-disable no-undef */
import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { fakeItem } from '../../../lib/testUtils';
import Product from '../Product';
import { CartStateProvider } from '../../../lib/cartState';

const product = fakeItem();

describe('<Product />', () => {
  it('matches the default snapshot', () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider>
          <Product product={product} />
        </MockedProvider>
      </CartStateProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
