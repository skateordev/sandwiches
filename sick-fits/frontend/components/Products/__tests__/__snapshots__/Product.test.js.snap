// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> matches the default snapshot 1`] = `
<div>
  <div
    class="ItemStyles-sc-16pk14u-0 BOSXe"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <h3
      class="Title-sc-16nq74k-0 ifVtXh"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="PriceTag-sc-nwbk6t-0 hxPUFn"
    >
      €50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit 📝
      </a>
      <button
        aria-busy="false"
        class="AddToCart__AddToCartStyled-sc-6bl65o-0 fHiDfE"
        type="button"
      >
        Add
         to cart 🛒
      </button>
    </div>
  </div>
</div>
`;
